src/promise.rs

```diff
@@ -0,0 +1,171 @@
+use crate::support::Opaque;
+use crate::Context;
+use crate::Function;
+use crate::Local;
+use crate::Value;
+
+extern "C" {
+  fn v8__Promise__Resolver__New(context: *mut Context) -> *mut PromiseResolver;
+  fn v8__Promise__Resolver__GetPromise(
+    resolver: *mut PromiseResolver,
+  ) -> *mut Promise;
+  fn v8__Promise__Resolver__Resolve(
+    resolver: *mut PromiseResolver,
+    context: *mut Context,
+    value: *mut Value,
+  ) -> bool;
+  fn v8__Promise__Resolver__Reject(
+    resolver: *mut PromiseResolver,
+    context: *mut Context,
+    value: *mut Value,
+  ) -> bool;
+  fn v8__Promise__State(promise: *mut Promise) -> PromiseState;
+  fn v8__Promise__HasHandler(promise: *mut Promise) -> bool;
+  fn v8__Promise__Result(promise: *mut Promise) -> *mut Value;
+  fn v8__Promise__Catch(
+    promise: *mut Promise,
+    context: *mut Context,
+    handler: *mut Function,
+  ) -> *mut Promise;
+  fn v8__Promise__Then(
+    promise: *mut Promise,
+    context: *mut Context,
+    handler: *mut Function,
+  ) -> *mut Promise;
+  fn v8__Promise__Then2(
+    promise: *mut Promise,
+    context: *mut Context,
+    on_fulfilled: *mut Function,
+    on_rejected: *mut Function,
+  ) -> *mut Promise;
+}
+
+#[derive(Debug, PartialEq)]
+#[repr(C)]
+pub enum PromiseState {
+  Pending,
+  Fulfilled,
+  Rejected,
+}
+
+#[repr(C)]
+pub struct Promise(Opaque);
+
+/// An instance of the built-in Promise constructor (ES6 draft).
+impl Promise {
+  /// Returns the value of the [[PromiseState]] field.
+  pub fn state(&mut self) -> PromiseState {
+    unsafe { v8__Promise__State(&mut *self) }
+  }
+
+  /// Returns true if the promise has at least one derived promise, and
+  /// therefore resolve/reject handlers (including default handler).
+  pub fn has_handler(&mut self) -> bool {
+    unsafe { v8__Promise__HasHandler(&mut *self) }
+  }
+
+  /// Returns the content of the [[PromiseResult]] field. The Promise must not
+  /// be pending.
+  pub fn result<'sc>(&mut self) -> Local<'sc, Value> {
```

Comment:
There are several methods on Promise that return a Local handle, such as this one.
With local handles you have to be careful that their lifetime is properly bound to the HandleScope in which they live.
This function doesn't do that. `'sc` is effectively an unbound lifetime here and thus the returned Local will happily outlive its HandleScope, which is unsafe.

You can fix this by adding a HandleScope parameter to the function signature (even if the argument is unused).
So:
`pub fn result<'sc>(&mut self, _scope: &mut HandleScope<'sc>) -> Local<'sc, Value> {`